src/server/server.py

```diff
@@ -2073,6 +2086,10 @@ def emit_environmental_data(**params):
         'aux_voltage': [data['voltage'] for data in INTERFACE.ina219_data],
         'aux_current': [data['current'] for data in INTERFACE.ina219_data],
         'aux_power': [data['power'] for data in INTERFACE.ina219_data],
+        'aux_voltage0' : [data['voltage0'] for data in INTERFACE.ads1x15_data],
```

Comment:
I think this should all be added to aux_voltage, else we will end up with a zoo of keys. Entries probably need to become json objects with some sort of device identifer property and the voltage value. In fact, this should probably be done consistency over voltage, current, power, temperate, ...

Reply:
Ok. I will be adding the key 'GUI_LABELS' in the config.json file. Do you agree config is a good place for these labels as they depend on what the user has wired? For example, mine has two batteries and a schottky diode, so I have labels of  `["Supply [v]","Battery 1 [v]","Battery 2 [v]"]`.

One issue I have with the dynamic sensor detection is that it's possible to have multiple ADS voltage sensors or multiple INA power sensors unless the config file has an explicit hex address for each set of data parameters.

Here are the current options I can think of:

1. Limit config to only one sensor of each type. The address is auto-detected, and the config matches with just that sensor.

2. Allow config to support multiple sensors by formatting them as a list in config.json, but require a user to specify the hex address for each sensor. Dynamic allocation of sensors will then not be needed.

3. Use the same config settings for all sensors of the same type. 

What are your thoughts?

Reply:
I would just have a 'label' key rather than specifically tie it to the gui by mentioning gui. For example, the label might be exposed over json REST API in the future.

I think (2) is the best, in fact, we can still do auto-detection of sensors, and for any sensor found see if there is any explicit config for it (based on address) and use that. If there is no config, we can generate a default label based on the address.

(I would also assume ppl may have a combination of INA219 and ADS, and whatever else rather than one or the other).

Reply:
Got it. 

To support different configurations on the same sensor same, should they be formatted as 
1) A list within the each sensor config
```
"ADS1X15": {[
                {
		"HARDWARE_NAME": "ADS1115",
		"CONNECTED_CHANNELS": [0,1],
		"GAINS": [1,1],
		"R1_VALUES": [22,22],
		"R2_VALUES": [3.3,3.3],
		"CORRECTION_FACTORS": [1.00,1.00]
                "ADDRESS": False
                 },
                 {
		"HARDWARE_NAME": "ADS1115",
		"CONNECTED_CHANNELS": [0,1],
		"GAINS": [1,1],
		"R1_VALUES": [22,22],
		"R2_VALUES": [3.3,3.3],
		"CORRECTION_FACTORS": [1.00,1.00]
                "ADDRESS": 0x48
                  {
                 

},
```
or
2) Separate keys with the address if specified in the name otherwise default like so:
```
"ADS1X15": {
		"HARDWARE_NAME": "ADS1115",
		"CONNECTED_CHANNELS": [0,1],
		"GAINS": [1,1],
		"R1_VALUES": [22,22],
		"R2_VALUES": [3.3,3.3],
		"CORRECTION_FACTORS": [1.00,1.00]
},
"ADS1X15_0x48"{
		"HARDWARE_NAME": "ADS1115",
		"CONNECTED_CHANNELS": [0,1],
		"GAINS": [1,1],
		"R1_VALUES": [22,22],
		"R2_VALUES": [3.3,3.3],
		"CORRECTION_FACTORS": [1.00,1.00]
}

```

While (1) is more clean, it would be more confusing for users to set up that don't have any coding background.

Reply:
I would do it like (2), but just have the address as the key. And maybe wrap the whole lot under a top-level 'devices'/'sensors' key. I wouldn't worry about being able to specify generic defaults. If a sensor is auto-detected, but has no explicit config then we should fallback to some safe defaults and if that is not practical, skip/ignore that sensor and log some message saying config is required for said sensor. INA219 library can use/auto-detect safe defaults, but it looks like ADS requires some config, so maybe a log message makes more sense in its case.

Reply:
Ok. I'll go with that. Please note I have to make all the keys different to be valid JSON so that's why I put the key in it. They are all in sensors and with fully unique hex keys in both the name and then the decimal key as an address. 

Sound good on the rest of it. I'll print if there's errors in import.

Reply:
How about if we have the address as the top level key, then an array with the config for each individual sensor at that address?

Reply:
Oh so instead of naming it ADS1X15/0x48 it's just config data? Seems simple enough.

Reply:
Give me a couple of mins and I'll paste what i think it should look like.

Reply:
Done. Although the key is redundant with the address, I think it's fine because later if there are multiple possible sensors on the same hex address, it will still support that.

Reply:
```
		"0x4B": {
// for ads, config is array entry per sensor at the same address in channel order
			 [{
			"label": "Supply [v]",
			"gain": 1,
			"r1_value": 22,
			"r2_value": 3.3,
			"correction_factor": "1.0065"
			},
			 {
			"label": "Battery 1 [v]",
			"gain": 1,
			"r1_value": 22,
			"r2_value": 3.3,
			"correction_factor": "1.0127175"
			},
			 {
			"label": "Battery 2 [v]",
			"gain": 1,
			"r1_value": 22,
			"r2_value": 3.3,
			"correction_factor": "1.02739"
			}]
		},
		"0x40": {
// for ina, config is map entry (label) per sensor at the same address
			"VOLTAGE_LABEL": "Voltage",
			"CURRENT_LABEL": "Current",
			"POWER_LABEL": "Power",
		},



```

Reply:
Labels are required to be unique per measurement type (then we can group associated measurement types by name).

Reply:
The environment data might be better organised by label, measurement type rather than measurement type, label.
```
"Battery 2 [V]": {"voltage": 432},
"My INA": {"voltage": 12, "current": 0.7, "power: "1.2"}
```
vs.
```
"voltage": {"Battery 2 [V]": 432, "My INA": 12},
"current": {"My INA": 0.7},
"power": {"My INA": 1.2},
```
It is easy enough to get from one to the other, don't know which you think is best. By label first might be easier for the UI to consume.

Reply:
I want to go by label. Right now as the code is written, the sensors are set up by address as the key. 
The sensors get initialized by the address key which works well and prevents you trying to add multiple sensors to the same key. It's also very explicit on which sensors get which config values.

Later when the measurements are taken, the label gets combined with the measurement. Any formatting for significant digits can be done there. Then, this data can all be bundled up in an array with the label as the key and a string value of the formatted measurement.

Thus, the UI can iterate through and no longer cares what kind of sensor took the environmental data.

I'll implement this and see how it works. It should allow both flexibility in the sensor configuration while also eliminating all UI custom code dependent on the sensor.